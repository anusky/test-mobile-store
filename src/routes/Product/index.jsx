import { Form, useRouteLoaderData } from "react-router-dom";
import { isEmpty } from "src/utils";
import { PAGE_STRUCTURE, PRODUCT_ATTRIBUTES } from "src/utils/constants";
import { getProductById } from "src/utils/productManagement";

import { addProductToCart } from "src/utils/cartManagement";
import Selector from "src/components/Selector";

/* istanbul ignore next */
export const productLoader = async ({ params }) => {
  return getProductById(params.productId);
};

/* istanbul ignore next */
export const productAction = async ({ request }) => {
  const formData = await request.formData();
  const colorCode = formData.get(PRODUCT_ATTRIBUTES.colorCode);
  const storageCode = formData.get(PRODUCT_ATTRIBUTES.storageCode);
  const id = formData.get(PRODUCT_ATTRIBUTES.id);

  const res = await addProductToCart({ colorCode, storageCode, id });
  if (res.error) throw res;
  return { colorCode, storageCode, id };
};

/* istanbul ignore next */
export const productHandle = {
  crumb: (data) => ({ content: `${data.brand}, ${data.model}` }),
};

const Product = () => {
  const productInfo = useRouteLoaderData(PAGE_STRUCTURE.product.clientPageId);

  return (
    <div
      data-testid="product-page-component"
      className="container mx-auto grid lg:grid-cols-2 py-8 px-8"
    >
      <div>
        <img
          className="mx-auto object-contain"
          src={productInfo.imgUrl}
          alt={`${productInfo.brand}, ${productInfo.model}`}
          height={200}
          width={150}
        />
      </div>
      <div className="grid gap-y-4">
        <h2>General Description</h2>
        <ul className="product-information grid">
          <li>
            <h3>Brand:</h3> <span>{productInfo.brand}</span>
          </li>
          <li>
            <h3>Model:</h3> <span>{productInfo.model}</span>
          </li>
          <li>
            <h3>Price:</h3>
            <span>
              {isEmpty(productInfo.price) ? "-" : `${productInfo.price}â‚¬`}
            </span>
          </li>
          <li>
            <h3>CPU:</h3> <span>{productInfo.cpu}</span>
          </li>
          <li>
            <h3>RAM:</h3> <span>{productInfo.ram}</span>
          </li>
          <li>
            <h3>OS:</h3> <span>{productInfo.os}</span>
          </li>
          <li>
            <h3>Display Res:</h3> <span>{productInfo.displayResolution}</span>
          </li>
          <li>
            <h3>Battery:</h3> <span>{productInfo.battery}</span>
          </li>
          <li>
            <h3>Cameras:</h3>
            <div className="grid px-2">
              {productInfo?.primaryCamera && (
                <span>Primary: {`${productInfo?.primaryCamera}`}</span>
              )}
              {productInfo?.secondaryCmera && (
                <span>Secondary: {`${productInfo?.secondaryCmera}`}</span>
              )}
            </div>
          </li>
          <li>
            <h3>Dimentions:</h3> <span>{productInfo.dimentions}</span>
          </li>
          <li>
            <h3>Weight:</h3> <span>{productInfo.weight} grams</span>
          </li>
        </ul>

        <Form
          className="grid gap-y-4"
          method="post"
          action={`/${productInfo.id}`}
        >
          <h2>Make your selections:</h2>
          <input
            type="text"
            name={PRODUCT_ATTRIBUTES.id}
            defaultValue={productInfo.id}
            hidden
          />
          <Selector
            name={PRODUCT_ATTRIBUTES.colorCode}
            title="Select a color:"
            options={productInfo.options.colors}
          />
          <Selector
            name={PRODUCT_ATTRIBUTES.storageCode}
            title="Select a storage:"
            options={productInfo.options.storages}
          />
          <button
            type="submit"
            className="rounded-full px-4 bg-gradient-to-r from-indigo-900 to-slate-900 text-white w-fit py-2"
          >
            Add to cart
          </button>
        </Form>
      </div>
    </div>
  );
};

export default Product;
