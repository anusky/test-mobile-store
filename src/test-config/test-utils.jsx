import "./__mocks__/intersectionObserverMock";

import { render } from "@testing-library/react";
import { createMemoryRouter, RouterProvider } from "react-router-dom";
import PropTypes from "prop-types";

import productInfo from "./__fixtures__/product.json";
import productList from "./__fixtures__/productList.json";
import cart from "./__fixtures__/cart.json";
import { ProductContextProvider } from "src/context/productContext";
import { PAGE_STRUCTURE } from "src/utils/constants";

const customRender = (ui, options = {}) => {
  const { routerInitialIndex = 0 } = options;

  const AllTheProviders = ({ children }) => {
    const routes = [
      {
        path: "/",
        element: children,
        id: PAGE_STRUCTURE.home.clientPageId,
        loader: () => ({
          productList,
          cart,
        }),
        handle: {
          crumb: () => ({ url: "/", content: "home route" }),
        },
        children: [
          {
            path: "/:productId",
            element: children,
            id: PAGE_STRUCTURE.product.clientPageId,
            loader: () => productInfo,
            handle: {
              crumb: () => ({ url: "/", content: "product route" }),
            },
          },
        ],
      },
    ];

    const router = createMemoryRouter(routes, {
      initialEntries: ["/", "/123"],
      initialIndex: routerInitialIndex,
    });
    return (
      <RouterProvider router={router}>
        <ProductContextProvider
          value={{
            list: productList,
            cart: cart,
          }}
        >
          {children}
        </ProductContextProvider>
      </RouterProvider>
    );
  };

  AllTheProviders.propTypes = {
    children: PropTypes.node.isRequired,
  };

  return render(ui, { wrapper: AllTheProviders, ...options });
};

export * from "@testing-library/react";
export { customRender as render };
