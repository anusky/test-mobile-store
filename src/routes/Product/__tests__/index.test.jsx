import { act, render, screen } from "src/test-config/test-utils";
import Product from "..";

describe("Product page coverage", () => {
  it("Product page renders properly", async () => {
    render(<Product />, { routerInitialIndex: 1 });
    expect(await screen.findByTestId("product-page-component")).toBeTruthy();
  });

  it("matches snapshot", async () => {
    let asFragment;
    await act(() => {
      const component = render(<Product />, { routerInitialIndex: 1 });

      asFragment = component.asFragment;
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
