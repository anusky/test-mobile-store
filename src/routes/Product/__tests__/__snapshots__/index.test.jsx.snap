// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product page coverage matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto grid lg:grid-cols-2 py-8 px-8"
    data-testid="product-page-component"
  >
    <div>
      <img
        alt="Acer, Iconia Talk S"
        class="mx-auto object-contain"
        height="200"
        src="https://front-test-api.herokuapp.com/images/ZmGrkLRPXOTpxsU4jjAcv.jpg"
        width="150"
      />
    </div>
    <div
      class="grid gap-y-4"
    >
      <h2>
        General Description
      </h2>
      <ul
        class="product-information grid"
      >
        <li>
          <h3>
            Brand:
          </h3>
           
          <span>
            Acer
          </span>
        </li>
        <li>
          <h3>
            Model:
          </h3>
           
          <span>
            Iconia Talk S
          </span>
        </li>
        <li>
          <h3>
            Price:
          </h3>
          <span>
            170â‚¬
          </span>
        </li>
        <li>
          <h3>
            CPU:
          </h3>
           
          <span>
            Quad-core 1.3 GHz Cortex-A53
          </span>
        </li>
        <li>
          <h3>
            RAM:
          </h3>
           
          <span>
            2 GB RAM
          </span>
        </li>
        <li>
          <h3>
            OS:
          </h3>
           
          <span>
            Android 6.0 (Marshmallow)
          </span>
        </li>
        <li>
          <h3>
            Display Res:
          </h3>
           
          <span>
            7.0 inches (~69.8% screen-to-body ratio)
          </span>
        </li>
        <li>
          <h3>
            Battery:
          </h3>
           
          <span>
            Non-removable Li-Ion 3400 mAh battery (12.92 Wh)
          </span>
        </li>
        <li>
          <h3>
            Cameras:
          </h3>
          <div
            class="grid px-2"
          >
            <span>
              Primary: 13 MP,autofocus
            </span>
            <span>
              Secondary: 2 MP,720p
            </span>
          </div>
        </li>
        <li>
          <h3>
            Dimentions:
          </h3>
           
          <span>
            191.7 x 101 x 9.4 mm (7.55 x 3.98 x 0.37 in)
          </span>
        </li>
        <li>
          <h3>
            Weight:
          </h3>
           
          <span>
            260 grams
          </span>
        </li>
      </ul>
      <form
        action="/ZmGrkLRPXOTpxsU4jjAcv"
        class="grid gap-y-4"
        method="post"
      >
        <h2>
          Make your selections:
        </h2>
        <input
          hidden=""
          name="id"
          type="text"
          value="ZmGrkLRPXOTpxsU4jjAcv"
        />
        <div
          class="inline-flex gap-x-4 relative"
        >
          <label>
            Select a color:
          </label>
          <select
            name="colorCode"
          >
            <option
              value="1000"
            >
              Black
            </option>
          </select>
        </div>
        <div
          class="inline-flex gap-x-4 relative"
        >
          <label>
            Select a storage:
          </label>
          <select
            name="storageCode"
          >
            <option
              value="2000"
            >
              16 GB
            </option>
            <option
              value="2001"
            >
              32 GB
            </option>
          </select>
        </div>
        <button
          class="rounded-full px-4 bg-gradient-to-r from-indigo-900 to-slate-900 text-white w-fit py-2"
          type="submit"
        >
          Add to cart
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
